---
title: "Using and installing the CTK pipeline using CLIPflexR"
author: "Kathryn Rozen-Gagnon, Ji-Dung Luo and Thomas Carroll"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{install_CLIPflexR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
baseDir <- "/Users/thomascarroll/Desktop/Projects/brc/clipRUTest/"
```

## CLIPflexR

The CLIPflexR package provides a set of its functions for the processing and analysis of differing forms of ClIP data for use within the R framework.

For more information on the processing of data using CLIPflexR you can see our [CTK](StandardandBrdU_Processing_CTK.html) and [CLIPflexR](Processing_to_matrix.html) vignettes

## Installation

### Installing CLIPflexR

We can install CLIPflexR package from Github using the devtools package:

```{r installCLIPflexR,echo=TRUE,eval=FALSE}
install.packages("devtools")
devtools::install_github("kathrynrozengagnon/CLIPflexR")
```

### Installing Rfastp... will be stable dependency once CLIPflexR is released.
We also can install the Rfastp package which may use to pre-process FastQ/FastA and de-duplicate BAM files as an alternative to the standard CTK/FASTX workflow.

If you are running <  R version 4.0, use  the following  lines to install the [Github Rfastp package](https://github.com/RockefellerUniversity/Rfastp):

```{r installrfastp_git,echo=TRUE,eval=FALSE}
devtools::install_github("RockefellerUniversity/rfastp")
```

If you are running >= R version 4.0, you can install the [Bioconductor development Rfastp package](https://bioconductor.org/packages/Rfastp/):

```{r installrfastp_BioC,echo=TRUE,eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')

BiocManager::install("Rfastp")
```

### Installing CTK software requirements.

The CTK toolkit requires many external softwares to be available on the system path for use within their pipeline scripts. 

One solution to the meet these system requirements is to create a Conda environment containing all the required software. 

The CondaSysReq package provides a simple workflow to install self-contained Conda environments associated to packages using the Reticulate library. 

This approach also allows us to capture the current Conda environment using the same tools we capture R library versions by means of the Renv package.

First we need to install the [Herper package](https://github.com/RockefellerUniversity/Herper).

### Install Herper:
If you are running >=  R version 4.0:
```{r installHerper4,echo=TRUE,eval=FALSE}
library(devtools)
install_github("RockefellerUniversity/Herper")
```

If you are running <  R version 4.0:
```{r installHerper3.5,echo=TRUE,eval=FALSE}
library(devtools)
install_github("RockefellerUniversity/Herper@3.5")
```

We can now use the Herper package to create the required Conda environment and requirments for the CLIPflexR package.

The below code will install the required external software to a Conda environment within the default location (same path as the Reticulate package's Conda environments). 

The path to the Conda executable and the name of the Conda environment:

```{r installConda_Reqs,echo=TRUE,eval=FALSE}
library(Herper)
CondaInfo <- install_CondaSysReqs("CLIPflexR")
CondaInfo$pathToConda
CondaInfo$environment
```

The installed Conda environment may be accessed out of R just as with a standard environment.

Executables can be found the environment's bin directory. Here we check the directory contains the fastx and homer executable.

```{r lookAtEnv,echo=FALSE,eval=FALSE,engine='bash'}
ls /Users/thomascarroll/Library/r-miniconda/envs/clipr*/bin/ \
| grep -E -- 'fastx|homer' | head -n 20
```

```{r lookAtEnv2,echo=FALSE,eval=FALSE}
dir("/Users/thomascarroll/Library/r-miniconda/envs/clipR_0.1.17/bin/",pattern="fastx|homer")
```

### Installing CTK software.

The [**ctk toolkit**](https://zhanglab.c2b2.columbia.edu/index.php/CTK_Documentation) is installed as part of the **CLIPflexR** **install_ctk()** function.


```{r installCTK,echo=TRUE,eval=FALSE}
CLIPflexR::install_ctk()
```

This function installs CTK and it's perl library dependencies to a user specified directory or if a Conda environment has been created with Herper to within the same Conda Environment associated to the package. 


```{r lookAtEnv3,echo=FALSE,eval=FALSE}
dir("/Users/thomascarroll/Library/r-miniconda/envs/clipR_0.1.17/bin/ctk",pattern="")
```


### Using the CTK pipeline in CLIPflexR.

Once we have installed the requirements for CTK we can make use of the tools within CLIPflexR very easily.

First we load the CLIPflexR package and review the loading messages informing us of any established conda paths.

```{r CheckPaths,echo=TRUE,eval=TRUE}
library(CLIPflexR)
```

We can also list the important paths from the environmental options set by CLIPflexR.

```{r CheckPaths2,echo=TRUE,eval=TRUE}
getOption("CLIPflexR.condaEnv")
getOption("CLIPflexR.ctk")
getOption("CLIPflexR.czplib")
```

Now that these are installed, all functions with the CLIPflexR package will use these by default without further configuration of paths or environmental variables.

```{r testa,echo=TRUE,eval=TRUE}
require(CLIPflexR)
Fox3_Std <- system.file("extdata/Fox3_Std_small.fq.gz",package="CLIPflexR")
Fox3_Std_filtered <- ctk_fastqFilter(Fox3_Std,
                                     outFile = "SRR1107535_Test.fastq",
                                     qsFilter="mean:0-29:20",verbose=TRUE)

```
