---
title: "Processing reads, mapping, and building a count matrix"
author: "Kathryn Rozen-Gagnon, Ji-Dung Luo and Thomas Carroll"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{processing_to_matrix}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
baseDir <- "/Users/kathryn/Reprocess_all_paper_datasets/test_processing_vignette"
```

## Process standard/BrdU-CLIP libraries using CLIPflexR

This is an alternative to the CTK-only processing pipeline for CLIP data, and returns a peak matrix with genomic locations of peaks and counts in each peak by sample

##Initial processing and quality control

### Decompress fastq files
# [documentation](fastq_quality_filter.html)

```{r decompress,echo=TRUE,eval=FALSE}
#define the path to your compressed fastq file
testFQ <- list.files(baseDir, full.names = TRUE)
testFQ
FqFile <- decompress(testFQ,overwrite=TRUE)
FqFile
```

### You can look at the quality of your data  **(wrapping FASTX's quality stats)**
# [documentation](fastx_quality_stats.html)

```{r quality_stats,echo=TRUE,eval=FALSE}
#defaults are set for illumina quality encoding, minimum quality score of 20 for 80% of bases
stats <- fastx_quality_stats(FqFile)
stats <- read.delim(stats,header = T, sep = "\t")
head(stats)
```

### Quality filter **(wrapping FastX's quality filter)**
# [documentation](fastq_quality_filter.html)

```{r quality_filter,echo=TRUE,eval=FALSE}
#defaults for all wrapped FASTX tools are set for illumina quality encoding. You neeed too  minimum quality score of 20 for 80% of bases
FqFile_QF <- fastq_quality_filter(FqFile)
FqFile_QF
```

### Collapse exact PCR duplicates **(wrapping FASTX's quality stats)**
# [documentation](fastq_quality_stats.html)

```{r collapse_PCR_duplicates,echo=TRUE,eval=FALSE}
coll_FqFile_QF <- fastx_collapser(FqFile_QF)
coll_FqFile_QF
```

##Follow these steps to process standard CLIP libraries generated by ligation  of 5'  and 3' linkers to RNAs

### First split up your samples if they are multiplexed **(wrapping FASTX's barcode splitter)**
# [documentation](fastx_barcode_splitter.html)

```{r split_samples,echo=TRUE,eval=FALSE}
#provide the path to your barcode file,  which must contain one sample name and it's corresponding barcode sequence for  each line, and be tab delimited

#For example:
#Aag2_rIgG_Exp6A CGATG
#Aag2_rIgG_Exp6B TTAGG

BCfile <- dir(baseDir, pattern="BC",  full.names = TRUE)
BCfile

split_samples <- fastx_barcode_splitter(coll_FqFile_QF, BCfile, verbose = TRUE)

```


We can now use the CondaSysReqs package to create the required Conda environment and requirments for the CLIPflexR package.

The below code will install the required external software to a Conda environment within the default location (same path as the Reticulate package's Conda environments). 

The path to the Conda executable and the name of the Conda environment 
```{r installConda_Reqs,echo=TRUE,eval=FALSE}
library(CondaSysReqs)
CondaInfo <- install_CondaSysReqs("CLIPflexR")
CondaInfo$pathToConda
CondaInfo$environment
```

The installed Conda environment may be accessed out of R just as with a standard environment.

Executables can be found the environment's bin directory. Here we check the directory contains the fastx and homer executable.

```{r lookAtEnv,echo=FALSE,eval=FALSE,engine='bash'}
ls /Users/thomascarroll/Library/r-miniconda/envs/clipr*/bin/ \
| grep -E -- 'fastx|homer' | head -n 20
```

```{r lookAtEnv2,echo=FALSE,eval=FALSE}
dir("/Users/thomascarroll/Library/r-miniconda/envs/clipR_0.1.17/bin/",pattern="fastx|homer")
```

### Installing CTK software.

The [**ctk toolkit**](https://zhanglab.c2b2.columbia.edu/index.php/CTK_Documentation) is installed as part of the **CLIPflexR** **install_ctk()** function.


```{r installCTK,echo=TRUE,eval=FALSE}
CLIPflexR::install_ctk()
```

This function installs CTK and it's perl library dependencies to a user specified directory or if a Conda environment has been created with CondaSysReqs to within the same Conda Environment associated to the package. 


```{r lookAtEnv3,echo=FALSE,eval=FALSE}
dir("/Users/thomascarroll/Library/r-miniconda/envs/clipR_0.1.17/bin/ctk",pattern="")
```


### Using the CTK pipeline in CLIPflexR.

Once we have installed the requirements for CTK we can make use of the tools within CLIPflexR very easily.

First we load the CLIPflexR package and review the loading messages informing us of any established conda paths.

```{r CheckPaths,echo=TRUE,eval=TRUE}
library(CLIPflexR)
```

We can also list the important paths from the environmental options set by CLIPflexR.


```{r CheckPaths2,echo=TRUE,eval=TRUE}
getOption("CLIPflexR.condaEnv")
getOption("CLIPflexR.ctk")
getOption("CLIPflexR.czplib")
```


Now that these are installed, all functions with the CLIPflexR package will use these by default without further configuration of paths or environmental variables.





```{r testa,echo=TRUE,eval=TRUE}
require(CLIPflexR)
Fox3_Std <- system.file("extdata/Fox3_Std_small.fq.gz",package="CLIPflexR")
Fox3_Std_filtered <- ctk_fastqFilter(Fox3_Std,
                                     outFile = "SRR1107535_Test.fastq",
                                     qsFilter="mean:0-29:20",verbose=TRUE)

```
